### ISSUE TITLE
repeated errors in the netstats module (tcp_retrans_percentage and tcpext_tcploss_percentage metrics)

### ISSUE THREAD

**Issue by martinwalsh** [ganglia/gmond_python_modules#88]
Environment: 
- CentOS 6.2
- Python 2.6.6
- gmond 3.3.0

We're seeing the following error messages in the syslog every 15 seconds on several CentOS 6.2 systems. 

```
[PYTHON] Can't call the metric handler function for [tcpext_tcploss_percentage] in the python module [netstats].#012
[PYTHON] Can't call the metric handler function for [tcp_retrans_percentage] in the python module [netstats].#012
```

These are consistently  `tcpext_tcploss_percentage` and `tcp_retrans_percentage`, on all systems, and other metrics appear to be working fine. 

Thanks in advance!

**Comment by acutchin** [comment]
This is a realy PITA. Any updates on a solution?

**Comment by rubinlinux** [comment]
I am seeing this still. It is flooding the logs with:

[PYTHON] Can't call the metric handler function for [tcpext_listendrops] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [tcp_attemptfails] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [tcpext_tcploss_percentage] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [tcp_retrans_percentage] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [tcp_outsegs] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [tcp_insegs] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [udp_indatagrams] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [udp_outdatagrams] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [udp_inerrors] in the python module [netstats].
[PYTHON] Can't call the metric handler function for [udp_rcvbuferrors] in the python module [netstats].

...over and over

ganglia 3.6.0-7 (debian)

**Comment by refraction-ray** [comment]
As pointed by @shawn174 , the solution is this patch: https://github.com/ganglia/monitor-core/pull/123/ in ganglia/monitor-core. I will elaborate this issue a bit here just in case it is helpful for future viewers with the same issue.

The problem is because `/usr/lib/ganglia/python_modules/netstats.py` (typical installation path on ubuntu by apt) doesn't handle `ZeroDivisionError`. Therefore, if some metrics are unchanged in the last period, the denominator is just zero causing exception in the function. The solution is rather simple, just catch ZeroDivisionError in `get_tcploss_percentage` and `get_retrans_percentage` functions as shown in the above PR.

This fix is already merged as early as Oct 16, 2013. However, ganglia 3.6.0 release is even older than this, which is released on Apr 30, 2013. It is surprising to see that apt still provided such an older version of ganglia (at least it is still 3.6.0 on Ubuntu 18.04). Therefore, the easiest way is modifing `/usr/lib/ganglia/python_modules/netstats.py` directly, or one can install a newer version of ganglia instead of the default version provided by apt. 

To sum up, the repeated errors in syslog is caused by an old python script bug (not catching ZeroDivision) and a very old version of ganglia provided by some distribution (Ubuntu18.04 provided a release 18-13=5 years old!).

**Comment by ghost** [comment]
We are on 3.7.2 and still experiencing this problem. Maybe it was reintroduced in one of the last versions released. 

Will likely try editing netstats.py manually as a solution. Commenting this here though in case anyone else on 3.7.2 might encounter this

**Comment by ghost** [comment]
Still monitoring, but so far I believe I was able to implement a fix for 3.7.2

Manually changed /usr/lib64/ganglia/python_modules/netstats.py (your installation location may differ) to catch ZeroDivisionErrors in the **delta** function. I can't see how this would have a negative effect on monitoring since the exception should only invoke in cases of divide-by-zero situations. So far so good on my end on a node I'm testing this on.

```
def get_delta(name):
    """Return change over time for the requested metric"""

    # get metrics
    [curr_metrics, last_metrics] = get_metrics()

    parts = name.split("_")
    group = parts[0]
    metric = "_".join(parts[1:])

    try:
        delta = (float(curr_metrics['data'][group][metric]) - float(last_metrics['data'][group][metric])) / (curr_metrics['time'] - last_metrics['time'])
        if delta < 0:
            print name + " is less 0"
            delta = 0
    except KeyError:
        delta = 0.0
    except ZeroDivisionError:
        delta = 0.0

    return delta
```

The only change being the addition of the lines:

```
except ZeroDivisionError:
        delta = 0.0
```
